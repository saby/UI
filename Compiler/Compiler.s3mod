<?xml version="1.0" encoding="UTF-8" ?>
<ui_module id="1456f36f-76ef-4323-a2dd-a22bc91e7209" name="Compiler" package="Платформа" responsible="Крылов М. А." required="1" version="1.00">
  <description>
    Интерфесный модуль компилятора шаблонов WML фреймворка Wasaby
  </description>

  <depends/>

  <load_after/>

  <parameters/>

  <resources/>

  <preload/>

</ui_module>
