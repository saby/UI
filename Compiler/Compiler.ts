/**
 * @description Main wml compiler module.
 * @author Крылов М.А.
 */

import * as ComponentCollector from './core/deprecated/ComponentCollector';
import { parse } from './html/Parser';
import { createErrorHandler } from './utils/ErrorHandler';
import getWasabyTagDescription from './core/Tags';
import { traverse } from './core/bridge';
import * as codegenBridge from './codegen/bridge';
import * as templates from './codegen/templates';
import { ISource, Source } from './utils/Source';
import { IOptions, Options } from './utils/Options';
import { ModulePath } from './utils/ModulePath';
import { ITranslationKey } from './i18n/Dictionary';

/**
 * Флаг - генерировать rk-функции
 * @todo https://online.sbis.ru/opendoc.html?guid=ea8a25dd-5a2f-4330-8d6f-599c8c5878dd
 */
const USE_GENERATE_CODE_FOR_TRANSLATIONS = false;

/**
 * Represents compiler interface.
 */
export interface ICompiler {
   /**
    * Compile input text.
    * @param text Source code text.
    * @param options Compiler options.
    */
   compile(text: string, options: IOptions): Promise<IArtifact>;
}

/**
 * Represents artifact interface.
 */
export interface IArtifact {

   /**
    * Node name for require.
    */
   nodeName: string;

   /**
    * Array of happened errors.
    * FIXME: release error handler.
    */
   errors: Error[];

   /**
    * Compile result: Javascript source code.
    */
   text: string | null;

   /**
    * Translations dictionary.
    */
   localizedDictionary: ITranslationKey[];

   /**
    * Array of input file dependencies.
    */
   dependencies: string[];

   /**
    * Flag whether compile result is stable.
    * It is stable if and only if there were not fatal errors.
    */
   stable: boolean;
}

/**
 * Create empty artifact with default values.
 */
function createArtifact(options: IOptions): IArtifact {
   return {
      nodeName: ModulePath.createNodeName(options.modulePath),
      errors: [],
      text: null,
      localizedDictionary: [],
      dependencies: null,
      stable: false
   };
}

/**
 * Represents abstract syntax tree interface as array of "object" - abstract syntax nodes.
 * FIXME: release interfaces for these nodes.
 */
interface IAST extends Array<Object> {
   childrenStorage: string[];
   reactiveProps: string[];
   templateNames: string[];
   __newVersion: boolean;
   hasTranslations: boolean;
}

/**
 * Represents interface for traverse resulting object.
 */
interface ITraversed {

   /**
    * Abstract syntax tree.
    */
   ast: IAST;

   /**
    * Translations dictionary.
    */
   localizedDictionary: ITranslationKey[];

   /**
    * Array of input file dependencies.
    */
   dependencies: string[];

   /**
    * Collection of inline template names.
    */
   templateNames: string[];

   hasTranslations: boolean;
}

/**
 * Fix traverse result.
 * @param rawTraversed Actual traverse result.
 * @param dependencies Array of dependencies.
 */
function fixTraversed(rawTraversed: IAST | { astResult: IAST; words: ITranslationKey[]; }, dependencies: string[]): ITraversed {
   if (Array.isArray(rawTraversed)) {
      return {
         ast: rawTraversed as IAST,
         localizedDictionary: [],
         templateNames: rawTraversed.templateNames,
         hasTranslations: rawTraversed.hasTranslations,
         dependencies
      };
   }
   return {
      ast: rawTraversed.astResult as IAST,
      localizedDictionary: rawTraversed.words,
      templateNames: rawTraversed.astResult.templateNames,
      hasTranslations: rawTraversed.astResult.hasTranslations,
      dependencies
   };
}

/**
 * Represents base compiler methods for wml and tmpl.
 */
abstract class BaseCompiler implements ICompiler {
   protected constructor() { }

   /**
    * Do initialize before compilation process.
    */
   abstract initWorkspace(templateNames: string[]): void;

   /**
    * Clean needed variables after compilation process.
    */
   abstract cleanWorkspace(): void;

   /**
    * Create source for input source text.
    * @param text Input source text.
    * @param fileName Source file name.
    */
   abstract createSource(text: string, fileName: string): ISource;

   /**
    * Generate template module.
    * @param func Template function.
    * @param deps Array of dependencies.
    * @param reactive Array of names of reactive variables.
    * @param path Template module path.
    * @param hasTranslations Translation unit contains translation constructions.
    */
   abstract generateModule(func: any, deps: string[], reactive: string[], path: ModulePath, hasTranslations: boolean): string;

   /**
    * Generate code for template.
    * @param traversed Traverse object.
    * @param options Compiler options.
    */
   generate(traversed: ITraversed, options: IOptions): string {
      const codeGenOptions = {
         ...options,
         generateTranslations: (
             options.generateCodeForTranslations && USE_GENERATE_CODE_FOR_TRANSLATIONS
             || !USE_GENERATE_CODE_FOR_TRANSLATIONS
         ) && traversed.hasTranslations
      };
      // tslint:disable:prefer-const
      let tmplFunc = codegenBridge.getFunction(traversed.ast, null, codeGenOptions, null);
      if (!tmplFunc) {
         throw new Error('Шаблон не может быть построен. Не загружены зависимости.');
      }
      return this.generateModule(
          tmplFunc, traversed.dependencies, traversed.ast.reactiveProps, options.modulePath, traversed.hasTranslations
      );
   }

   /**
    * Traverse source code.
    * @param source Source code.
    * @param options Compiler options.
    */
   traverse(source: ISource, options: IOptions): Promise<ITraversed> {
      return new Promise<ITraversed>((resolve: any, reject: any) => {
         try {
            // TODO: реализовать whitespace visitor и убрать флаг needPreprocess
            const needPreprocess = options.modulePath.extension === 'wml';
            const errorHandler = createErrorHandler(!options.fromBuilderTmpl);
            // tslint:disable:prefer-const
            let parsed = parse(source.text, options.fileName, {
               xml: true,
               allowComments: true,
               allowCDATA: true,
               compatibleTreeStructure: true,
               rudeWhiteSpaceCleaning: true,
               normalizeLineFeed: true,
               cleanWhiteSpaces: true,
               needPreprocess,
               tagDescriptor: getWasabyTagDescription,
               errorHandler
            });
            const hasFailures = errorHandler.hasFailures();
            const lastMessage = errorHandler.popLastErrorMessage();
            errorHandler.flush();
            if (hasFailures) {
               reject(new Error(lastMessage));
               return;
            }
            const dependencies = ComponentCollector.getComponents(parsed);
            // tslint:disable:prefer-const
            let traversed = traverse(parsed, options);
            traversed.addCallbacks(
               (rawTraversed) => resolve(fixTraversed(rawTraversed, dependencies)),
               (error) => reject(error)
            );
         } catch (error) {
            reject(error);
         }
      });
   }

   /**
    * Compile input source code into Javascript code.
    * @param text Input source code.
    * @param options Compiler options.
    */
   compile(text: string, options: IOptions): Promise<IArtifact> {
      return new Promise<IArtifact>((resolve: any, reject: any) => {
         let artifact: IArtifact = createArtifact(options);
         try {
            const source: ISource = this.createSource(text, options.fileName);
            this.traverse(source, options)
               .then((traversed) => {
                  try {
                     this.initWorkspace(traversed.templateNames);
                     artifact.text = this.generate(traversed, options);
                     artifact.localizedDictionary = traversed.localizedDictionary;
                     artifact.dependencies = traversed.dependencies;
                     artifact.stable = true;
                     resolve(artifact);
                  } catch (error) {
                     artifact.errors.push(error);
                     reject(artifact);
                  } finally {
                     this.cleanWorkspace();
                  }
               })
               .catch((error) => {
                  this.cleanWorkspace();
                  artifact.errors.push(error);
                  reject(artifact);
               });
         } catch (error) {
            artifact.errors.push(error);
            reject(artifact);
         }
      });
   }
}

/**
 * This class represents methods to compile tmpl files.
 */
class CompilerTmpl extends BaseCompiler {
   constructor() {
      super();
   }

   /**
    * Do initialize before compilation process.
    */
   initWorkspace(templateNames: string[]): void {
      codegenBridge.initWorkspaceTMPL(templateNames);
   }

   /**
    * Clean needed variables after compilation process.
    */
   cleanWorkspace(): void {
      codegenBridge.cleanWorkspace();
   }

   /**
    * Create source for input source text.
    * @param text Input source text.
    * @param fileName Source file name.
    */
   createSource(text: string, fileName: string): ISource {
      return new Source(text, fileName);
   }

   /**
    * Generate template module.
    * @param func Template function.
    * @param deps Array of dependencies.
    * @param reactive Array of names of reactive variables.
    * @param path Template module path.
    * @param hasTranslations Translation unit contains translation constructions.
    */
   generateModule(func: any, deps: string[], reactive: string[], path: ModulePath, hasTranslations: boolean): string {
      return templates.generateTmplDefine(
         path.module, path.extension, func, deps, reactive, hasTranslations
      );
   }
}

/**
 * This class represents methods to compile wml files.
 */
class CompilerWml extends BaseCompiler {
   constructor() {
      super();
   }

   /**
    * Do initialize before compilation process.
    */
   initWorkspace(templateNames: string[]): void {
      codegenBridge.initWorkspaceWML(templateNames);
   }

   /**
    * Clean needed variables after compilation process.
    */
   cleanWorkspace(): void {
      codegenBridge.cleanWorkspace();
   }

   /**
    * Create source for input source text.
    * @param text Input source text.
    * @param fileName Source file name.
    */
   createSource(text: string, fileName: string): ISource {
      return new Source(text, fileName);
   }

   /**
    * Generate template module.
    * @param func Template function.
    * @param deps Array of dependencies.
    * @param reactive Array of names of reactive variables.
    * @param path Template module path.
    * @param hasTranslations Translation unit contains translation constructions.
    */
   generateModule(func: any, deps: string[], reactive: string[], path: ModulePath, hasTranslations: boolean): string {
      const module = templates.generateDefine(
         path.module, path.extension, func, deps, reactive, hasTranslations
      );
      return templates.clearSourceFromDeprecated(module);
   }
}

let DoT;

/**
 * This class represents methods to compile xhtml files.
 */
class CompilerXHTML implements ICompiler {
   /**
    * Generate template module.
    * @param func Template function.
    * @param path Template module path.
    */
   generate(func: any, path: ModulePath): string {
      const localizationModule = 'i18n!' + path.getInterfaceModule();
      const templateModuleRequire = 'html!' + path.module;
      const template = func.toString().replace(/[\n\r]/g, '');
      return 'define("' + templateModuleRequire + '",["' + localizationModule + '"],function(){' +
         'var f=' + template + ';' +
         'f.toJSON=function(){' +
         'return {$serialized$:"func", module:"' + templateModuleRequire + '"}' +
         '};return f;});';
   }

   /**
    * Compile input source code into Javascript code.
    * @param text Input source code.
    * @param options Compiler options.
    */
   compile(text: string, options: IOptions): Promise<IArtifact> {
      return new Promise((resolve: any, reject: any) => {
         let artifact: IArtifact = createArtifact(options);
         if (!DoT) {
            DoT = requirejs.defined('Core/js-template-doT') && requirejs('Core/js-template-doT');
         }
         try {
            // tslint:disable:prefer-const
            let config = DoT.getSettings();
            // tslint:disable:prefer-const
            let template = DoT.template(text, config, undefined, undefined, options.modulePath.module);
            artifact.text = this.generate(template, options.modulePath);
            artifact.stable = true;
            resolve(artifact);
         } catch (error) {
            artifact.errors.push(error);
            reject(artifact);
         }
      });
   }
}

/**
 * This class only represents returning error.
 */
class ErrorCompiler implements ICompiler {
   compile(text: string, options: IOptions): Promise<IArtifact> {
      let artifact = createArtifact(options);
      artifact.errors.push(new Error(
         'Данное расширение шаблона не поддерживается. Получен шаблон с расширением "' + options.modulePath.extension + '". Ожидалось одно из следующих расширений: wml, tmpl, xhtml.'
      ));
      return Promise.reject(artifact);
   }
}

/**
 * Get actual compiler with input source file extension.
 * @param extension Source file extension.
 */
function getCompiler(extension: string): ICompiler {
   switch (extension) {
      case 'wml':
         return new CompilerWml();
      case 'tmpl':
         return new CompilerTmpl();
      case 'xhtml':
         return new CompilerXHTML();
      default:
         return new ErrorCompiler();
   }
}

/**
 * Common class for compiling.
 */
export class Compiler implements ICompiler {

   /**
    * Compile input source code into Javascript code.
    * @param text Input source code.
    * @param config Compiler options.
    */
   compile(text: string, config: IOptions): Promise<IArtifact> {
      const options = new Options(config);
      const compiler = getCompiler(options.modulePath.extension);
      return compiler.compile(text, options);
   }
}
