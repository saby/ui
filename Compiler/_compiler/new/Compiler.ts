/**
 * @author Krylov M.A.
 */

import type { ICompiler, IOptions, IArtifact } from '../Interface';
import type { IErrorHandler } from '../utils/ErrorHandler';
import type { ISource } from '../utils/Source';
import type { Node } from '../html/Nodes';
import type { Ast } from '../core/Ast';
import type { IAnnotation } from '../core/IRAnnotator';
import type { IDescription } from '../irCodegen/Interface';

import { ModulePath } from '../utils/ModulePath';
import { createErrorHandler } from '../utils/ErrorHandler';
import Scope from '../core/Scope';
import { Source } from '../utils/Source';
import { parse } from '../html/Parser';
import getWasabyTagDescription from '../core/Tags';
import { Parser } from '../expressions/Parser';
import { createTextTranslator } from '../i18n/Translator';
import traverse from '../core/Traverse';
import { annotate } from '../core/IRAnnotator';
import createModuleProcessor from '../irCodegen/Module';
import createGenerator from '../irCodegen/Generator';

import {
    ENTRY_POINT_IDENTIFIER,
    TRANSLATION_FUNCTION,
    DEBUG_FUNCTION,
    IS_WINDOW_UNDEFINED_IDENTIFIER,
    STRINGS_IDENTIFIER
} from '../irCodegen/Constants';

/**
 * Флаг - генерировать rk-функции
 * @todo https://online.sbis.ru/opendoc.html?guid=ea8a25dd-5a2f-4330-8d6f-599c8c5878dd
 */
const USE_GENERATE_CODE_FOR_TRANSLATIONS = false;

/**
 * Create empty artifact with default values.
 */
function createArtifact(options: IOptions): IArtifact {
    return {
        nodeName: ModulePath.createNodeName(options.modulePath),
        errors: [],
        text: null,
        umdText: null,
        localizedDictionary: [],
        dependencies: null,
        stable: false,
    };
}

// FIXME: Разобраться с пробелами в текстовых узлах
function shouldPreprocessTextNodes(options: IOptions): boolean {
    const uiModuleName = options.modulePath.getInterfaceModule();
    if (uiModuleName === 'HelpFilling') {
        return false;
    }
    return options.modulePath.extension === 'wml';
}

function createSource(text: string, fileName: string): ISource {
    return new Source(text, fileName);
}

function checkErrors(errorHandler: IErrorHandler) {
    const hasFailures = errorHandler.hasFailures();

    if (hasFailures) {
        const lastMessage = hasFailures ? errorHandler.popLastErrorMessage() : undefined;
        errorHandler.flush();

        throw new Error(lastMessage);
    }
}

export class NewCompilerWml implements ICompiler {

    compile(text: string, options: IOptions): Promise<IArtifact> {
        return new Promise<IArtifact>(
            (resolve: (IArtifact) => void, reject: (IArtifact) => void) => {
                const artifact = this.compileSync(text, options);
                if (artifact.stable) {
                    resolve(artifact);
                    return;
                }
                reject(artifact);
            }
        );
    }

    compileSync(text: string, options: IOptions): IArtifact {
        const artifact: IArtifact = createArtifact(options);
        const scope = new Scope(!options.fromBuilderTmpl);
        const errorHandler = createErrorHandler(!options.fromBuilderTmpl);
        const source: ISource = createSource(text, options.fileName);

        try {
            const html = this.parse(source, options, errorHandler);
            const ast = this.traverse(html, options, scope, errorHandler);
            const annotation = this.annotate(ast, options, errorHandler);
            const description = this.generate(annotation, options, errorHandler);

            for (const moduleType of options.moduleType) {
                switch (moduleType) {
                    case 'amd':
                        artifact.text = this.wrapWithModule(
                            description,
                            options,
                            moduleType
                        );
                        break;

                    case 'umd':
                        artifact.umdText = this.wrapWithModule(
                            description,
                            options,
                            moduleType
                        );
                        break;
                }
            }

            artifact.dependencies = description.dependencies;
            artifact.localizedDictionary = scope.getTranslationKeys();

            if (!artifact.text && !artifact.umdText) {
                artifact.errors.push(
                    new Error(
                        'Внутреняя ошибка компилятора: ' +
                        `Шаблон не был скомпилирован ни в один из форматов ${JSON.stringify(
                            options.moduleType
                        )}`
                    )
                );
            }

            artifact.stable = artifact.errors.length === 0;
        } catch (error) {
            artifact.errors.push(error);
        } finally {
            errorHandler.flush();
        }

        return artifact;
    }

    private parse(source: ISource, options: IOptions, errorHandler: IErrorHandler): Node[] {
        // TODO: реализовать whitespace visitor и убрать флаг needPreprocess
        const html = parse(source.text, options.fileName, {
            xml: true,
            allowComments: true,
            allowCDATA: true,
            compatibleTreeStructure: true,
            rudeWhiteSpaceCleaning: true,
            normalizeLineFeed: true,
            cleanWhiteSpaces: true,
            needPreprocess: shouldPreprocessTextNodes(options),
            tagDescriptor: getWasabyTagDescription,
            errorHandler
        });

        checkErrors(errorHandler);

        return html;
    }

    private traverse(parsed: Node[], options: IOptions, scope: Scope, errorHandler: IErrorHandler): Ast[] {
        const config = {
            expressionParser: new Parser(),
            hierarchicalKeys: true,
            errorHandler,
            allowComments: false,
            textTranslator: createTextTranslator(options.componentsProperties || { }),
            generateTranslations:
                (USE_GENERATE_CODE_FOR_TRANSLATIONS && !!options.generateCodeForTranslations) ||
                !USE_GENERATE_CODE_FOR_TRANSLATIONS,
            hasExternalInlineTemplates: options.hasExternalInlineTemplates,
            checkInlineTemplateName: options.isWasabyTemplate
        };
        const traverseOptions = {
            fileName: options.fileName,
            scope,
            translateText: true
        };

        const ast = traverse(parsed, config, traverseOptions);

        checkErrors(errorHandler);

        return ast;
    }

    private annotate(ast: Ast[], options: IOptions, errorHandler: IErrorHandler): IAnnotation {
        const moduleName = ModulePath.createNodeName(options.modulePath);
        const annotation = annotate(errorHandler, ast, moduleName);

        checkErrors(errorHandler);

        return annotation;
    }

    private generate(annotation: IAnnotation, options: IOptions, errorHandler: IErrorHandler): IDescription<string> {
        const generator = createGenerator(options.ESVersion, options.isReleaseMode, options.isWasabyTemplate, 1);
        const description = generator.generate(annotation);

        checkErrors(errorHandler);

        return description;
    }

    private wrapWithModule(description: IDescription<string>, options: IOptions, moduleType: string) {
        const processor = createModuleProcessor(moduleType);
        processor.setModuleName(description.moduleName);
        processor.setStrictMode(true);

        // обязательные именованные зависимости добавляются в самом начале
        processor.addDependency('Compiler/IR', ENTRY_POINT_IDENTIFIER);
        const translationFunctionName = options.isWasabyTemplate ? TRANSLATION_FUNCTION : `_${TRANSLATION_FUNCTION}`;
        if (description.injections.varTranslate) {
            processor.addDependency(`i18n!${options.modulePath.getInterfaceModule()}`, translationFunctionName);
        }

        // следом добавляем все остальные зависимости шаблона
        for (let index = 0; index < description.dependencies.length; ++index) {
            processor.addDependency(description.dependencies[index]);
        }

        // добавляем определения используемых в шаблоне сущностей
        if (description.injections.funcDebug) {
            processor.addCodeBlock(`function ${DEBUG_FUNCTION}() { debugger; }`);
        }

        if (description.injections.funcDefaultContextGetter) {
            processor.addCodeBlock(description.injections.funcDefaultContextGetter);
        }

        if (description.injections.varIsWindowUndefined) {
            processor.addCodeBlock(`var ${IS_WINDOW_UNDEFINED_IDENTIFIER} = typeof window === "undefined";`);
        }

        if (description.injections.varStrings) {
            processor.addCodeBlock(`var ${STRINGS_IDENTIFIER} = ${description.injections.varStrings};`);
        }

        if (description.injections.varTranslate && !options.isWasabyTemplate) {
            processor.addCodeBlock(
                `function ${TRANSLATION_FUNCTION}(key, context, pluralNumber) { return ${translationFunctionName}(key, context, pluralNumber); }`
            );
        }

        processor.setReturnableExport(description.exports);

        return processor.compile();
    }
}
