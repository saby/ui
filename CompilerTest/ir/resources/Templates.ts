import * as Element from 'wml!CompilerTest/ir/resources/Element';
import * as If from 'wml!CompilerTest/ir/resources/If';
import * as For from 'wml!CompilerTest/ir/resources/For';
import * as Foreach from 'wml!CompilerTest/ir/resources/Foreach';
import * as InlineTemplate from 'wml!CompilerTest/ir/resources/InlineTemplate';
import * as ContentOption from 'wml!CompilerTest/ir/resources/ContentOption';
import * as StaticPartial from 'wml!CompilerTest/ir/resources/StaticPartial';
import * as DynamicPartial from 'wml!CompilerTest/ir/resources/DynamicPartial';
import * as DynamicInline from 'wml!CompilerTest/ir/resources/DynamicInline';

export default {
    Element,
    If,
    For,
    Foreach,
    InlineTemplate,
    ContentOption,
    StaticPartial,
    DynamicPartial,
    DynamicInline
};
