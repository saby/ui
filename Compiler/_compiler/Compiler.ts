/**
 * @description Main wml compiler module.
 */

import type { IOptions, ICompiler, IArtifact } from './Interface';

import * as ComponentCollector from './core/deprecated/ComponentCollector';
import { parse } from './html/Parser';
import { createErrorHandler } from './utils/ErrorHandler';
import getWasabyTagDescription from './core/Tags';
import { traverseSync } from './core/bridge';
import * as codegenBridge from './codegen/bridge';
import { IMainTemplateFunction } from './codegen/Template';
import { clearSourceFromTmplBlocks } from './codegen/JsTemplates';
import { ISource, Source } from './utils/Source';
import { Options } from './utils/Options';
import { ModulePath } from './utils/ModulePath';
import { ITranslationKey } from './i18n/Dictionary';
import createModuleProcessor from './codegen/Module';
import { writeModuleBody } from './codegen/module/Body';
import { createGenerator } from './codegen/ECMAScript';

import { NewCompilerWml } from './new/Compiler';

/**
 * Ручка для управления новой кодогенерацией.
 */
const USE_NEW_CODE_GENERATION = true;

/**
 * Флаг - генерировать rk-функции
 * @todo https://online.sbis.ru/opendoc.html?guid=ea8a25dd-5a2f-4330-8d6f-599c8c5878dd
 */
const USE_GENERATE_CODE_FOR_TRANSLATIONS = false;

// Паттерн подстановок генерации кода вида /*#паттерн#*/.
const COMMENT_LIKE_SUBSTITUTION_PATTERN = /\/\*#[A-Z_0-9]+#\*\//;

// FIXME: Разобраться с пробелами в текстовых узлах
function shouldPreprocessTextNodes(options: IOptions): boolean {
    const uiModuleName = options.modulePath.getInterfaceModule();
    if (uiModuleName === 'HelpFilling') {
        return false;
    }
    return options.modulePath.extension === 'wml';
}

/**
 * Create empty artifact with default values.
 */
function createArtifact(options: IOptions): IArtifact {
    return {
        nodeName: ModulePath.createNodeName(options.modulePath),
        errors: [],
        text: null,
        umdText: null,
        localizedDictionary: [],
        dependencies: null,
        stable: false,
    };
}

/**
 * Represents abstract syntax tree interface as array of "object" - abstract syntax nodes.
 * FIXME: release interfaces for these nodes.
 */
interface IAST extends Array<Object> {
    childrenStorage: string[];
    reactiveProps: string[];
    templateNames: string[];
    __newVersion: boolean;
    hasTranslations: boolean;
}

/**
 * Represents interface for traverse resulting object.
 */
interface ITraversed {
    /**
     * Abstract syntax tree.
     */
    ast: IAST;

    /**
     * Translations dictionary.
     */
    localizedDictionary: ITranslationKey[];

    /**
     * Array of input file dependencies.
     */
    dependencies: string[];

    /**
     * Collection of inline template names.
     */
    templateNames: string[];

    /**
     * Flag that indicates that template contains translations.
     */
    hasTranslations: boolean;
}

/**
 * Check compiled JS code.
 * @param text {string} Compiled JS code.
 */
function validateCompiledText(text: string): void {
    if (COMMENT_LIKE_SUBSTITUTION_PATTERN.test(text)) {
        const firstPattern = text.match(COMMENT_LIKE_SUBSTITUTION_PATTERN);
        throw new Error(
            `Внутреняя ошибка компилятора: Обнаружена необработанная подстановка ${firstPattern[0]}`
        );
    }
}

/**
 * Represents base compiler methods for wml and tmpl.
 */
abstract class BaseCompiler implements ICompiler {
    /**
     * Do initialize before compilation process.
     */
    abstract initWorkspace(templateNames: string[]): void;

    /**
     * Clean needed variables after compilation process.
     */
    abstract cleanWorkspace(): void;

    /**
     * Create source for input source text.
     * @param text Input source text.
     * @param fileName Source file name.
     */
    abstract createSource(text: string, fileName: string): ISource;

    /**
     * Generate template module.
     * @param template Template function.
     * @param traversed Traversed object.
     * @param options Compiler options.
     * @param moduleType Module type in which template must be compiled.
     */
    generateModule(
        template: IMainTemplateFunction,
        traversed: ITraversed,
        options: IOptions,
        moduleType: string
    ): string {
        const processor = createModuleProcessor(moduleType);
        const generator = createGenerator(options.ESVersion);

        processor.setStrictMode(true);

        writeModuleBody(
            processor,
            template,
            traversed.dependencies,
            traversed.ast.reactiveProps,
            traversed.hasTranslations,
            options,
            generator
        );

        let source = processor.compile();

        if (options.isWasabyTemplate) {
            source = clearSourceFromTmplBlocks(source);
        }

        validateCompiledText(source);

        return source;
    }

    /**
     * Generate code for template.
     * @param traversed Traverse object.
     * @param options Compiler options.
     */
    generateFunction(traversed: ITraversed, options: IOptions): IMainTemplateFunction {
        const codeGenOptions = {
            ...options,
            generateTranslations:
                ((options.generateCodeForTranslations && USE_GENERATE_CODE_FOR_TRANSLATIONS) ||
                    !USE_GENERATE_CODE_FOR_TRANSLATIONS) &&
                traversed.hasTranslations,
        };
        const tmplFunc = codegenBridge.getFunction(traversed.ast, null, codeGenOptions, null, true);

        if (!tmplFunc) {
            throw new Error('Шаблон не может быть построен. Не загружены зависимости.');
        }

        return tmplFunc;
    }

    /**
     * Traverse source code.
     * @param source Source code.
     * @param options Compiler options.
     */
    traverse(source: ISource, options: IOptions): ITraversed {
        // TODO: реализовать whitespace visitor и убрать флаг needPreprocess
        const errorHandler = createErrorHandler(!options.fromBuilderTmpl);
        const parsed = parse(source.text, options.fileName, {
            xml: true,
            allowComments: true,
            allowCDATA: true,
            compatibleTreeStructure: true,
            rudeWhiteSpaceCleaning: true,
            normalizeLineFeed: true,
            cleanWhiteSpaces: true,
            needPreprocess: shouldPreprocessTextNodes(options),
            tagDescriptor: getWasabyTagDescription,
            errorHandler,
        });
        const hasFailures = errorHandler.hasFailures();
        const lastMessage = hasFailures ? errorHandler.popLastErrorMessage() : undefined;
        errorHandler.flush();
        if (hasFailures) {
            throw new Error(lastMessage);
        }
        const dependencies = ComponentCollector.getComponents(parsed);
        return traverseSync(parsed, options, dependencies);
    }

    /**
     * Compile input source code into Javascript code.
     * @param text Input source code.
     * @param options Compiler options.
     */
    compile(text: string, options: IOptions): Promise<IArtifact> {
        return new Promise<IArtifact>(
            (resolve: (IArtifact) => void, reject: (IArtifact) => void) => {
                const artifact = this.compileSync(text, options);
                if (artifact.stable) {
                    resolve(artifact);
                    return;
                }
                reject(artifact);
            }
        );
    }

    compileSync(text: string, options: IOptions): IArtifact {
        const artifact: IArtifact = createArtifact(options);
        try {
            const source: ISource = this.createSource(text, options.fileName);
            const traversed = this.traverse(source, options);

            this.initWorkspace(traversed.templateNames);
            const tmplFunc = this.generateFunction(traversed, options);

            for (const moduleType of options.moduleType) {
                switch (moduleType) {
                    case 'amd':
                        artifact.text = this.generateModule(
                            tmplFunc,
                            traversed,
                            options,
                            moduleType
                        );
                        break;

                    case 'umd':
                        artifact.umdText = this.generateModule(
                            tmplFunc,
                            traversed,
                            options,
                            moduleType
                        );
                        break;
                }
            }

            artifact.localizedDictionary = traversed.localizedDictionary;
            artifact.dependencies = traversed.dependencies;

            if (!artifact.text && !artifact.umdText) {
                artifact.errors.push(
                    new Error(
                        'Внутреняя ошибка компилятора: ' +
                            `Шаблон не был скомпилирован ни в один из форматов ${JSON.stringify(
                                options.moduleType
                            )}`
                    )
                );
            }

            artifact.stable = artifact.errors.length === 0;
        } catch (error) {
            artifact.errors.push(error);
        } finally {
            this.cleanWorkspace();
        }
        return artifact;
    }
}

/**
 * This class represents methods to compile tmpl files.
 */
class CompilerTmpl extends BaseCompiler {
    /**
     * Do initialize before compilation process.
     */
    initWorkspace(templateNames: string[]): void {
        codegenBridge.initWorkspaceTMPL(templateNames);
    }

    /**
     * Clean needed variables after compilation process.
     */
    cleanWorkspace(): void {
        codegenBridge.cleanWorkspace();
    }

    /**
     * Create source for input source text.
     * @param text Input source text.
     * @param fileName Source file name.
     */
    createSource(text: string, fileName: string): ISource {
        return new Source(text, fileName);
    }
}

/**
 * This class represents methods to compile wml files.
 */
class CompilerWml extends BaseCompiler {
    /**
     * Do initialize before compilation process.
     */
    initWorkspace(templateNames: string[]): void {
        codegenBridge.initWorkspaceWML(templateNames);
    }

    /**
     * Clean needed variables after compilation process.
     */
    cleanWorkspace(): void {
        codegenBridge.cleanWorkspace();
    }

    /**
     * Create source for input source text.
     * @param text Input source text.
     * @param fileName Source file name.
     */
    createSource(text: string, fileName: string): ISource {
        return new Source(text, fileName);
    }
}

/**
 * This class only represents returning error.
 */
class ErrorCompiler implements ICompiler {
    compile(text: string, options: IOptions): Promise<IArtifact> {
        return Promise.reject(this.compileSync(text, options));
    }

    compileSync(text: string, options: IOptions): IArtifact {
        const artifact = createArtifact(options);
        artifact.errors.push(
            new Error(
                'Данное расширение шаблона не поддерживается. Получен шаблон с расширением "' +
                    options.modulePath.extension +
                    '". Ожидалось одно из следующих расширений: wml, tmpl.'
            )
        );
        return artifact;
    }
}

/**
 * Get actual compiler with input source file extension.
 * @param extension Source file extension.
 * @param useNewCodegen Use new code generation compiler.
 */
function getCompiler(extension: string, useNewCodegen: boolean): ICompiler {
    switch (extension) {
        case 'wml':
            if (useNewCodegen) {
                return new NewCompilerWml();
            }

            return new CompilerWml();

        case 'tmpl':
            return new CompilerTmpl();

        default:
            return new ErrorCompiler();
    }
}

/**
 * Common class for compiling.
 */
export class Compiler implements ICompiler {
    /**
     * Compile input source code into Javascript code.
     * @param text Input source code.
     * @param config Compiler options.
     */
    compile(text: string, config: IOptions): Promise<IArtifact> {
        const options = new Options(config);
        const compiler = getCompiler(
            options.modulePath.extension,
            // В requirejs плагинах на время тестирования этот параметр будет false
            options.useNewCodegen !== false && USE_NEW_CODE_GENERATION
        );
        return compiler.compile(text, options);
    }

    /**
     * Compile input source code into Javascript code.
     * @param text Input source code.
     * @param config Compiler options.
     */
    compileSync(text: string, config: IOptions): IArtifact {
        const options = new Options(config);
        const compiler = getCompiler(
            options.modulePath.extension,
            // В requirejs плагинах на время тестирования этот параметр будет false
            options.useNewCodegen !== false && USE_NEW_CODE_GENERATION
        );
        return compiler.compileSync(text, options);
    }
}
