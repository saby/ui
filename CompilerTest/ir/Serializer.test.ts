/* eslint-disable @typescript-eslint/no-magic-numbers */
import Templates from './resources/Templates';
import * as IR from 'Compiler/IR';

function invoke(fn: Function, data: object = {}): unknown[] {
    return fn(data, { }, { }, false);
}

describe('Compiler/_ir/Serializer', () => {
    it.skip('check content option', () => {
        const serialized = Templates.Serializable.templates[3].toJSON();
        const contentOption = IR.deserialize(serialized);

        expect(invoke(contentOption)).toMatchSnapshot();
    });
    it.skip('check serializer pattern', () => {
        const serialized = Templates.Serializable.templates[3].toJSON();
        expect(/^CONTENT_OPTION,(\d+),/.test(serialized)).toBeTruthy();
    });
});
