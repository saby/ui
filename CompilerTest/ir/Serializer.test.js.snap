// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiler/_ir/Serializer check content option 1`] = `"<div>first root content option</div><div>second root content option</div><div>third root content option</div>"`;
